opm/grid/GraphOfGridWrappers.hpp

```diff
@@ -0,0 +1,270 @@
+                      const GraphOfGrid<Dune::CpGrid>& gog)
+{
+    GraphOfGrid<Dune::CpGrid>* pGraph = const_cast<GraphOfGrid<Dune::CpGrid>*>(&gog);
+    Zoltan_Set_Num_Obj_Fn(zz, getGraphOfGridNumVertices, pGraph);
+    Zoltan_Set_Obj_List_Fn(zz, getGraphOfGridVerticesList, pGraph);
+    Zoltan_Set_Num_Edges_Multi_Fn(zz, getGraphOfGridNumEdges, pGraph);
+    Zoltan_Set_Edge_List_Multi_Fn(zz, getGraphOfGridEdgeList, pGraph);
+}
+
+/// \brief Adds well to the GraphOfGrid
+///
+/// Translates wells' cartesian ID to global ID used in the graph.
+/// Adding the well contracts vertices of the well into one vertex.
+///
+/// checkWellIntersections==true makes the algorithm check if wells
+/// intersect and if their cell IDs are present in the graph.
+/// Setting it to false makes the algorithm faster but leaves user
+/// responsible for keeping wells disjoint.
+void addFutureConnectionWells (GraphOfGrid<Dune::CpGrid>& gog,
+    const std::unordered_map<std::string, std::set<int>>& wells,
+    bool checkWellIntersections=true)
+{
+    // create compressed lookup from cartesian.
+    const auto& grid = gog.getGrid();
+    const auto& cpgdim = grid.logicalCartesianSize();
+    std::vector<int> cartesian_to_compressed(cpgdim[0]*cpgdim[1]*cpgdim[2], -1);
+    for( int i=0; i < grid.numCells(); ++i )
+    {
+        cartesian_to_compressed[grid.globalCell()[i]] = i;
+    }
+
+    for (const auto& w: wells)
+    {
+        std::set<int> wellsgID;
+        for (const int& cell : w.second)
+        {
+            int gID = cartesian_to_compressed.at(cell);
+            assert(gID!=-1); // well should be an active cell
+            wellsgID.insert(gID);
+        }
+        gog.addWell(wellsgID,checkWellIntersections);
+    }
+}
+
+/// \brief Add well cells' global IDs to the list
+///
+/// Output of the partitioning is missing vertices that were contracted.
+/// This function fills in omitted gIDs and gives them the properties
+/// (like process number and ownership) of their representative cell (well ID).
+template<typename TheTuple>
+void extendImportExportList (const GraphOfGrid<Dune::CpGrid>& gog,
+                                       std::vector<TheTuple>& cellList)
+{
+    // using TheTuple = std::tuple<int,int,char>; or std::tuple<int,int,char,int>
+    using CellList = std::vector<TheTuple>;
+    // make a list of wells for easy identification. Contains ID, begin, end
+    using iter = std::set<int>::const_iterator;
+    std::list<std::tuple<int,iter,iter>> wellList;
+    for (const auto& well : gog.getWells())
+    {
+        wellList.push_front(std::make_tuple(*well.begin(),well.begin(),well.end()));
+    }
+
+    CellList addToList;
+    // iterate once through the original cellList
+    for (const auto& cell : cellList)
+    {
+        // if a cell is in any well, add cells of the well to cellList
+        // and remove the well from the wellList to quicken the search
+        for (auto wID=wellList.begin(); wID!=wellList.end(); )
```

Comment:
It is hard to see and understand what is what below. Maybe using  a structured binding might help?
```
const auto& [ firstCell, begin, end] = *wID;